### PR TITLE
Distribution aware shape builder (WIP)

### PR DESCRIPTION
**Pull request by acunniffe**
https://www.loom.com/share/0f33dd3e102b41908f469a9cd3028ac7

### CODE REVIEW

File: core/optic/shared/src/main/scala/com/useoptic/diff/initial/DistributionAwareShapeBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+    val jsonLikeTraverser = new JsonLikeTraverser(RfcState.empty, visitor)
+
+    bodies.foreach(body => jsonLikeTraverser.traverse(Some(body), JsonTrail(Seq.empty), None))
+
+    aggregator
+  }
+
+
+  def toShapes(implicit trailValues: TrailValueMap, seed: String = s"${Random.alphanumeric take 6 mkString}"): ShapesToMake = {
```
Comment (devdoshi) [inline core/optic/shared/src/main/scala/com/useoptic/diff/initial/DistributionAwareShapeBuilder.scala:100]: We should inject the ID generator this time

File: core/optic/shared/src/main/scala/com/useoptic/diff/interactions/interpreters/MissingValueInterpreter.scala [hunk 2]
```diff
@@ -18,7 +18,8 @@ class MissingValueInterpreter(rfcState: RfcState) extends InteractiveDiffInterpr
   private val basicInterpretations = new BasicInterpretations(rfcState)
   private val descriptionInterpreters = new DiffDescriptionInterpreters(rfcState)
 
-  override def interpret(diff: InteractionDiffResult, interaction: HttpInteraction): Seq[InteractiveDiffInterpretation] = {
+  override def interpret(diff: InteractionDiffResult, interactions: Vector[HttpInteraction]): Seq[InteractiveDiffInterpretation] = {
+    val interaction = interactions.head
```
Comment (devdoshi) [inline core/optic/shared/src/main/scala/com/useoptic/diff/interactions/interpreters/MissingValueInterpreter.scala:22]: I think we could have kept this interface intact and added a new one that takes multiple interactions instead

File: core/optic/shared/src/main/scala/com/useoptic/diff/initial/DistributionAwareShapeBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+  def toMap: Map[JsonTrail, Vector[Option[JsonLike]]] = _internal.toMap
+
+  def valuesForTrail(trail: JsonTrail): Vector[Option[JsonLike]] = _internal.getOrElse(trail, Vector.empty)
+
+  def hasTrail(trail: JsonTrail) = _internal.contains(trail)
+
+}
+
+class ShapeBuilderVisitor(aggregator: TrailValueMap) extends JsonLikeVisitors {
```
Comment (devdoshi) [inline core/optic/shared/src/main/scala/com/useoptic/diff/initial/DistributionAwareShapeBuilder.scala:227]: We should use the JsonLikeAndSpecVisitors from now on - let me know if you want a hand migrating

Comment (acunniffe) [reply core/optic/shared/src/main/scala/com/useoptic/diff/initial/DistributionAwareShapeBuilder.scala:227]: Let’s do it during our session tomorrow. Didn’t realize it was new!

**Comment by acunniffe** [comment]
In #146

**Pull request closed without merging**